File: library/src/main/java/com/github/daneko/android/iab/model/Product.java
package com.github.daneko.android.iab.model;

import fj.data.Option;

import lombok.AccessLevel;
import lombok.AllArgsConstructor;
import lombok.Getter;
import lombok.ToString;
import lombok.Value;
import lombok.experimental.Wither;

/**
 *
 */
@Value
@AllArgsConstructor(access = AccessLevel.PRIVATE)
@ToString(of = {"productBaseInfo", "skuDetails"})
public class Product {
    ProductBaseInfo productBaseInfo;
    SkuDetails skuDetails;

    @Wither
    Option<Purchase> purchaseInfo;

    @Wither
    Option<String> developerPayload;

    public static Product create(final ProductBaseInfo productInfo, final SkuDetails skuDetails) {
        return new Product(productInfo, skuDetails, Option.none(), Option.none());
    }

    @Getter(lazy = true)
    private final BillingType billingType = productBaseInfo.getType();

    @Getter(lazy = true)
    private final String productId = skuDetails.getProductId();

    @Getter(lazy = true)
    private final IabItemType iabItemType = getBillingType().getIabItemType();

    public boolean canBuy() {
        return purchaseInfo.isNone();
    }

    public boolean canConsume() {
        return getBillingType() == BillingType.CONSUMPTION && purchaseInfo.isSome();
    }
}

File: library/src/main/kotlin/com/github/daneko/android/iab/model/Product.kt
package com.github.daneko.android.iab.model

import fj.data.Option

/**
 *
 */

data class Product(
        val productBaseInfo: ProductBaseInfo,
        val skuDetails: SkuDetails,
        val purchaseInfo: Option<Purchase>,
        val developerPayload: Option<String>
) {
    companion object {
        @JvmStatic
        fun create(productBaseInfo: ProductBaseInfo, skuDetails: SkuDetails): Product {
            return Product(productBaseInfo, skuDetails, Option.none(), Option.none())
        }

    }

    val billingType: BillingType by lazy { productBaseInfo.type }

    val productId: String by lazy { skuDetails.getProductId() }

    val IabItemType: IabItemType by lazy { billingType.iabItemType }

    fun canBuy(): Boolean {
        return purchaseInfo.isNone
    }

    fun canConsume(): Boolean {
        return billingType == BillingType.CONSUMPTION && purchaseInfo.isSome
    }
}
